# Part 1. Equation of a Slime

How many late days are you using for this assignment?   
Zero Days

In [3]:
!pip3 install pandas
!pip3 install scikit-learn

In [4]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

## 1. Loading the dataset

In [5]:
# Using pandas load the dataset
raw_data = pd.read_csv("science_data_large.csv")
# Output the first 15 rows of the data
print(raw_data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(raw_data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [7]:
# Take the pandas dataset and split it into our features (X) and label (y)
raw_x = raw_data[['Temperature °C', 'Mols KCL']]
raw_y = raw_data[['Size nm^3']]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
train_x, test_x, train_y, test_y = train_test_split(raw_x, raw_y, test_size=0.1, random_state=42)
# For grading consistency use random_state=42 

## 3. Perform a Linear Regression

In [8]:
# Use sklearn to train a model on the training set
model_linear_regression = LinearRegression()
model_linear_regression.fit(train_x, train_y)

# Create a sample datapoint and predict the output of that sample with the trained model
sample = [[300, 900]]
print(model_linear_regression.predict(sample))

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("\nTrain Score:",model_linear_regression.score(train_x, train_y))
print("\nTest Score:",model_linear_regression.score(test_x, test_y))

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
intercept = model_linear_regression.intercept_[0] 
coefficients = model_linear_regression.coef_[0]

print("\nIntercept:", intercept)
print("\nCoefficients:", coefficients)


[[779878.00426539]]

Train Score: 0.8610082946223788

Test Score: 0.8552472077276095

Intercept: -409391.4795834075

Coefficients: [ 866.14641337 1032.69506649]


/Users/<redacted>/Library/Mobile Documents/com~apple~CloudDocs/College/Classes Spring '25/COMPSCI 383/assignment4/assignment-4-sklearn-for-machine-learning-<redacted>/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)  
$h(x)_{linear}$ = $-409391.5$ + $866.15*Temp$ + $1032.7*Mols$

Report on score and explain meaning:  
The $R^2$ for the training dataset $R^2_{train}=0.861$ and the test dataset is $R^2_{test}=0.855$, this indicates that the model is able to explain 86.1% of the variance in the training dataset and 85.5% of the variance in the test dataset, a good result for a linear model. Furthermore, the difference in the $R^2$ scores between the test and training is 0.006, this indicates that the model is not overfitting the data.

## 4. Use Cross Validation

In [9]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
from sklearn.model_selection import cross_val_score, KFold

kfold = KFold(n_splits=5, random_state=42, shuffle=True)

scores_linear = cross_val_score(model_linear_regression, raw_x, raw_y, cv=kfold)

# Report on their finding and their significance
print("Cross Validation Scores:", scores_linear)
print("Mean CV Score:", scores_linear.mean())
print("Standard Deviation of CV Score:", scores_linear.std())

Cross Validation Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean CV Score: 0.8597294202684646
Standard Deviation of CV Score: 0.01838773713930639


Write findings here:   
The cross validation scores provide a robust measure of our model’s performance across different subsets of the data. 
By using 5-fold cross validation with a fixed random state, we ensure reproducibility and a fair evaluation. For instance, 
the CV scores are [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101], the mean CV score is approximately 0.86 with a 
low standard deviation. This indicates that our linear regression model consistently explains about 86% of the variance in 
the slime size across different data splits. Such a high and consistent score suggests that our model generalizes well to new, unseen data.

## 5. Using Polynomial Regression

In [10]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2)
raw_x_poly = poly.fit_transform(raw_x)

feature_names = poly.get_feature_names_out(['Temperature °C', 'Mols KCL'])
print("Polynomial feature names:", feature_names)

# Split the polynomial features into train and test sets
train_x_poly, test_x_poly, train_y, test_y = train_test_split(raw_x_poly, raw_y, test_size=0.1, random_state=42)

# Train a new Linear Regression model on the polynomial features
poly_model = LinearRegression()
poly_model.fit(train_x_poly, train_y)
# Report on the metrics and output the resultant equation as you did in Part 3.

# Compute cross validation scores on the polynomial features
cv_scores_poly = cross_val_score(poly_model, raw_x_poly, raw_y, cv=kfold)
intercept_poly = poly_model.intercept_[0] 
coefficients_poly = poly_model.coef_[0]

print("\nPolynomial Regression CV Scores:", cv_scores_poly)
print("\nMean CV Score:", cv_scores_poly.mean())
print("\nStandard Deviation of CV Score:", cv_scores_poly.std())
print("\nIntercept:", intercept_poly)
print("\nCoefficients:", coefficients_poly)




Polynomial feature names: ['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']

Polynomial Regression CV Scores: [1. 1. 1. 1. 1.]

Mean CV Score: 1.0

Standard Deviation of CV Score: 0.0

Intercept: 2.047856105491519e-05

Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27198083e-07  1.26448818e-11
  2.00000000e+00  2.85714287e-02]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)  
$h(x)_{polynomial}$ = $2.047856105491519*10^{-05}$ + $12.00*Temp$ + $-1.272*Mols$ + $(1.26*10^{-11})*Temp^2$ + $2.00*Temp*Mols$ + $0.0286*Mols^2$ 

Report on the score and interpret:  
The polynomial regression model uses a transformed feature space that includes both linear and quadratic terms, along with an interaction term between Temperature and Mols KCL. The cross validation scores (mean ± standard deviation) indicate that the model consistently performs at a certain level across different splits, in this case the CV scores are 1 indicating a perfect fit of the data and a StDev of 0.0. The equation derived from the model provides a more nuanced exact description of how these variables influence the slime size, potentially capturing nonlinear relationships that a simple linear model might miss.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [17]:
# Load the dataset. Then train and evaluate the classification models.

ckd_data = pd.read_csv("ckd_feature_subset.csv")

print("Dataset columns:", ckd_data.columns)
target_column = 'Target_ckd'

# Separate the features (X) and the target (y)
X = ckd_data.drop(columns=[target_column])
y = ckd_data[target_column]

from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

# Set up a 5-fold cross validation configuration
kfold = KFold(n_splits=5, shuffle=True, random_state=42)

# Define the classification models + NN
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "Support Vector Machines": SVC(random_state=42),
    "k-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(random_state=42, max_iter=1000)
}

# Dictionary to store results
results = {}

# Evaluate each model using cross validation
for name, model in models.items():
    cv_scores = cross_val_score(model, X, y, cv=kfold, scoring='accuracy')
    results[name] = {"Mean Accuracy": cv_scores.mean(), "Std Accuracy": cv_scores.std()}

# Display the results in a table
results_df = pd.DataFrame(results).T
print("Classification Model Performance:")
print(results_df)

Dataset columns: Index(['age', 'bp', 'wbcc', 'appet_poor', 'appet_good', 'rbcc', 'Target_ckd'], dtype='object')
Classification Model Performance:
                         Mean Accuracy  Std Accuracy
Logistic Regression           0.856559      0.066269
Support Vector Machines       0.928172      0.047601
k-Nearest Neighbors           0.927957      0.052440
Neural Network                0.935054      0.040466


## Results and Conclusion for Classification Experiments

The following Classification models were ran and to gauge performance of the models cross validation was used:

- Logistic Regression ${[Mean_{accuracy}: 0.86, Std_{accuracy}: 0.066]}$
- Support Vector Machines ${[Mean_{accuracy}: 0.93, Std_{accuracy}: 0.048]}$
- k-Nearest Neighbors ${[Mean_{accuracy}: 0.93, Std_{accuracy}: 0.052]}$
- Neural Network ${[Mean_{accuracy}: 0.94, Std_{accuracy}: 0.040]}$

Based on these results the best performing classification was provided by neural networks (94%), followed by KNN(93%) and Support Vector Machines(93%); with Logistic regression(86%) performing the worst. All the classification models had a std of ≈ 0 indicating they perform well throughout all the data, however, the performance of the polynomial regression is better than these methods. 



In [18]:
# Experiments with Neural Network.

# Define different neural network configurations
nn_configs = {
    "Config 1": MLPClassifier(hidden_layer_sizes=(50,), activation='relu', random_state=42, max_iter=1000),
    "Config 2": MLPClassifier(hidden_layer_sizes=(100,), activation='tanh', random_state=42, max_iter=1000),
    "Config 3": MLPClassifier(hidden_layer_sizes=(50, 50), activation='relu', random_state=42, max_iter=1000),
    "Config 4": MLPClassifier(hidden_layer_sizes=(50, 50, 50), activation='relu', random_state=42, max_iter=1000),
    "Config 5": MLPClassifier(hidden_layer_sizes=(100, 50, 25), activation='relu', random_state=42, max_iter=1000),
    "Config 6": MLPClassifier(hidden_layer_sizes=(256, 128, 64, 32, 16, 8), activation='relu', random_state=42, max_iter=1000),
    "Config 7": MLPClassifier(hidden_layer_sizes=(512, 256, 128, 64, 32, 16, 8), activation='relu', random_state=42, max_iter=1000),
    "Config 8": MLPClassifier(hidden_layer_sizes=(1000, 500, 100), activation='relu', random_state=42, max_iter=1000),
}

nn_results = {}

# Evaluate each neural network configuration using cross validation
for name, model in nn_configs.items():
    cv_scores = cross_val_score(model, X, y, cv=kfold, scoring='accuracy')
    nn_results[name] = {"Mean Accuracy": cv_scores.mean(), "Std Accuracy": cv_scores.std()}

# Convert the neural network results to a DataFrame and Print
nn_results_df = pd.DataFrame(nn_results).T
print("Neural Network Configurations Performance:")
print(nn_results_df)

/Users/<redacted>/Library/Mobile Documents/com~apple~CloudDocs/College/Classes Spring '25/COMPSCI 383/assignment4/assignment-4-sklearn-for-machine-learning-<redacted>/.venv/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Mobile Documents/com~apple~CloudDocs/College/Classes Spring '25/COMPSCI 383/assignment4/assignment-4-sklearn-for-machine-learning-<redacted>/.venv/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network Configurations Performance:
          Mean Accuracy  Std Accuracy
Config 1       0.935054      0.040466
Config 2       0.941505      0.037327
Config 3       0.954624      0.043696
Config 4       0.954624      0.043696
Config 5       0.967527      0.035340
Config 6       0.967527      0.035340
Config 7       0.967527      0.035340
Config 8       0.967527      0.035340


## Results and Conclusion for Neural Network Experiments

8 configs were tested with varying no. of layers and nodes per layer, ReLu activation was used for all configs with max iters. capped at 1000 (more iters were tested but yielded no accuracy differences). Based on the results [output of above cell] performance increased with more no. of layers and increased no. of nodes from config 1 - config 5, however, after increasing no. of layers from 3 there were no improvements to accuracy performance. Therefore, the best config. is a NN with 3 hidden layers of size: (>100, >50, >25) yielding a performance of 97% accuracy with a std. of ≈ 0.03. 

